Framework/PythonInterface/plugins/algorithms/PeakMatching.py

```diff
@@ -0,0 +1,307 @@
+
+        secondary_energies = {}
+
+        all_energies = {}
+
+        for element in peak_data:
+            primary_energy = [float(x[1]) for x in list(peak_data[element]['Primary'].items())]
+            secondary_energy = [float(x[1]) for x in list(peak_data[element]['Secondary'].items())]
+            primary_trans = [x[0] for x in list(peak_data[element]['Primary'].items())]
+            secondary_trans = [x[0] for x in list(peak_data[element]['Secondary'].items())]
+
+            primary_energies[element] = dict(zip(primary_trans, primary_energy))
+            secondary_energies[element] = dict(zip(secondary_trans, secondary_energy))
+            all_energies[element] = dict(zip(primary_trans + secondary_trans,
+                                             primary_energy + secondary_energy))
+
+        peak_data = {}
+        peak_data['Primary energy'] = primary_energies
+        peak_data['Secondary energy'] = secondary_energies
+        peak_data['All energies'] = all_energies
+
+        return peak_data
+
+    def make_count_table(self,name,data):
+        table = CreateEmptyTableWorkspace(OutputWorkspace=name)
+
+        table.addColumn("str", "Element")
+        table.addColumn("int", "Likelyhood(au)")
+
+        counts = {}
+
+        for entry in data:
+            if not(entry['element'] in counts):
+                counts[entry['element']] = entry['Rating']
+            counts[entry['element']] += entry['Rating']
+
+        sorted_data_by_count = sorted(counts.items(),key=lambda x: x[1], reverse=True)
+
+        for match in sorted_data_by_count:
+            row = [str(match[0]),match[1]]
+            table.addRow(row)
+        return table
+
+    def get_input_peaks(self,table,centre_column_name,sigma_column_name):
+        '''DEFINING WIDTH OF PEAKS
+        Width input which will be used to determine the standard
+        deviation and FWHM which will be used as error for fitting.
+
+        From 'Gauss' function in origin:
+        y = y0 + (A/(w*sqrt(PI/2)))*exp(-2*((x-xc)/w)^2)
+        which has been used to fit all data to this point, where:
+            A>0
+            y0 = offset
+            xc = centre
+            w = width
+            A = area
+
+        FWHM = w(sqrt(ln4))
+        sigma = w/2
+
+        '''
+        table_dict = mtd[table].toDict()
+        peaks = []
+        sigma = []
+        for i in range(len(table_dict[centre_column_name])):
+            peaks.append(table_dict[centre_column_name][i])
+            sigma.append(table_dict[sigma_column_name][i])
+
+        return list(zip(peaks, sigma))
+
+    def get_matches(self,peak_data, input_peaks):
+        all_matches = []
+        for x in peak_data:
+            raw_data = peak_data[x]
+            matches = []
+            for peak, sigma in input_peaks:
+                for element in raw_data:
+                    for transition, energy in raw_data[element].items():
+
+                        if peak == energy:
```

Comment:
is there much difference in performance if you have a for loop that loops over the number of sigma?
```
for num_sigma in range(0,3)
    if peak >= (energy - num_sigma*sigma) and peak <= (energy + sigma*num_sigma):
                            data = {}
                            data['element'] = element
                            data['energy'] = energy
                            data['error'] = sigma*num_sigma
                            data['peak_centre'] = peak
                            data['transition'] = transition
                            data['diff'] = abs(peak - energy)
                            data['Rating'] = 4-num_sigma
                            matches.append(data)

```

Reply:
There isn't much of a performance difference if it is done this way.

Reply:
I think the above is easier to understand and is less repeated code. Can we change it please?